pkg/ingest/store/local.go

```diff
@@ -0,0 +1,165 @@
+	if storageURI.Scheme != "local" {
+		return path.ErrBadPattern
+	}
+	root := path.Join(storageURI.Host, storageURI.Path)
+	if err := l.verifyAbsPath(root); err != nil {
+		return err
+	}
+
+	var entries []*ObjectStoreEntry
+	// verify and use cache - location is stored in continuation token
+	if options.ContinuationToken != "" && strings.HasPrefix(options.ContinuationToken, l.cacheLocation) {
+		cacheData, err := os.ReadFile(options.ContinuationToken)
+		if err == nil {
+			err = json.Unmarshal(cacheData, &entries)
+			if err != nil {
+				entries = nil
+			} else {
+				l.mark.ContinuationToken = options.ContinuationToken
+			}
+		}
+	}
+
+	// if needed scan all entries to import and calc etag
+	if entries == nil {
+		if err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			key := filepath.ToSlash(p)
+			if key < options.After {
+				return nil
+			}
+			if !info.Mode().IsRegular() {
+				return nil
+			}
+
+			// calculate etag for the file
+			f, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer func() { _ = f.Close() }()
+			hash := md5.New() //nolint:gosec
+			_, err = io.Copy(hash, f)
+			if err != nil {
+				return err
+			}
+
+			addr := "local://" + key
+			relativePath, err := filepath.Rel(root, p)
+			if err != nil {
+				return err
+			}
+			etag := hex.EncodeToString(hash.Sum(nil))
+			ent := &ObjectStoreEntry{
+				FullKey:     key,
+				RelativeKey: filepath.ToSlash(relativePath),
+				Address:     addr,
+				ETag:        etag,
+				Mtime:       info.ModTime(),
+				Size:        info.Size(),
+			}
+			entries = append(entries, ent)
+			return nil
+		}); err != nil {
+			return err
+		}
+		sort.Slice(entries, func(i, j int) bool {
+			return entries[i].FullKey < entries[j].FullKey
+		})
+
+		// store entries to cache file
+		if l.cacheLocation != "" {
+			jsonData, err := json.Marshal(entries)
+			if err != nil {
+				return err
+			}
+			const dirPerm = 0o755
+			_ = os.MkdirAll(l.cacheLocation, dirPerm)
+			cacheName := filepath.Join(l.cacheLocation, nanoid.Must()+"-import.json")
+			const cachePerm = 0o644
+			if err := os.WriteFile(cacheName, jsonData, cachePerm); err != nil {
+				return err
+			}
+			l.mark.ContinuationToken = cacheName
+		}
+	}
+
+	// search start position base on Last key
+	startIndex := sort.Search(len(entries), func(i int) bool {
+		return entries[i].FullKey > options.After
+	})
+	for i := startIndex; i < len(entries); i++ {
+		ent := *entries[i]
+		l.mark.LastKey = ent.FullKey
+		if err := walkFn(ent); err != nil {
+			return err
+		}
+	}
+	// delete cache in case we completed the iteration
+	if l.mark.ContinuationToken != "" {
+		if err := os.Remove(l.mark.ContinuationToken); err != nil {
+			return err
+		}
+	}
+	l.mark = Mark{}
+	return nil
+}
+
+func (l *LocalWalker) verifyAbsPath(root string) error {
```

Comment:
1. Thought the `ingest` package should not be coupled to `block` adapter package. But I see we already coupled them in azure.
2. Reuse the func will be in a level where we will pass all required parameters and the caller will have to check for adapter package errors.
If we no longer use the walker outside catalog/controller we can move the functionality into each adapter and it will solve all the above. Can be done in different PR.
Let me know what you think?

Reply:
Yeah, certainly can be in a separate PR.